src/Build/BackEnd/Node/MSBuildClient.cs

```diff
@@ -0,0 +1,477 @@
+                _dllLocation,
+                "/nologo",
+                "/nodemode:8"
+            };
+
+            try
+            {
+                Process msbuildProcess = LaunchNode(_exeLocation, string.Join(" ", msBuildServerOptions),  ServerEnvironmentVariables);
+                CommunicationsUtilities.Trace("Server is launched.");
+            }
+            catch (Exception ex)
+            {
+                CommunicationsUtilities.Trace($"Failed to launch the msbuild server: {ex.Message}");
+                _exitResult.MSBuildClientExitType = MSBuildClientExitType.LaunchError;
+                return false;
+            }
+
+            return true;
+        }
+
+        private Process LaunchNode(string exeLocation, string msBuildServerArguments, Dictionary<string, string> serverEnvironmentVariables)
+        { 
+            ProcessStartInfo processStartInfo = new ProcessStartInfo
+            {
+                FileName = exeLocation,
+                Arguments = msBuildServerArguments,
+                UseShellExecute = false
+            };
+
+            // TODO: do we really need to start msbuild server with these variables, performance-wise and theoretically thinking?
+            // We are sending them in build command as well.
+            foreach (var entry in serverEnvironmentVariables)
+            {
+                processStartInfo.Environment[entry.Key] = entry.Value;
+            }
+
+            // We remove env MSBUILDRUNSERVERCLIENT that might be equal to 1, so we do not get an infinite recursion here. 
+            processStartInfo.Environment["MSBUILDRUNSERVERCLIENT"] = "0";
+
+            processStartInfo.CreateNoWindow = true;
+            processStartInfo.UseShellExecute = false;
+
+            return Process.Start(processStartInfo) ?? throw new InvalidOperationException("MSBuild server node failed to lunch");
+        }
+
+
+        private void SendBuildCommand(string commandLine, NamedPipeClientStream nodeStream)
+        {
+            ServerNodeBuildCommand buildCommand = GetServerNodeBuildCommand(commandLine);
+            WritePacket(_nodeStream, buildCommand);
+            CommunicationsUtilities.Trace("Build command send...");
+        }
+
+        private ServerNodeBuildCommand GetServerNodeBuildCommand(string commandLine)
+        {
+
+            Dictionary<string, string> envVars = new Dictionary<string, string>();
+
+            IDictionary environmentVariables = Environment.GetEnvironmentVariables();
+            foreach (var key in environmentVariables.Keys)
+            {
+                envVars[(string)key] = (string) (environmentVariables[key] ?? "");
+            }
+
+            foreach (var pair in ServerEnvironmentVariables)
+            {
+                envVars[pair.Key] = pair.Value;
+            }
+
+            // We remove env MSBUILDRUNSERVERCLIENT that might be equal to 1, so we do not get an infinite recursion here. 
```

Comment:
This env variable controls if we call the regular build execution or the client. It is processed in xmake and client. If we do not remove it and call any other process with propagation of this variable (node spawning processes, for example), they might run the client instead of doing what is needed. 
I want to keep processing of this variable close there and prevent such situations with certainty. We do not want to this variable propagate any further than current process. 
Also, I am thinking on options in somehow other way switch between client and regular build.